concent_api/core/payments/report.py

```diff
@@ -0,0 +1,202 @@
+
+
+class PairReport(NamedTuple):
+    requestor: Client
+    provider: Client
+    requestor_ethereum_addresses: list
+    provider_ethereum_addresses: list
+    pending_responses: PendingResponses
+    payments: Payments
+
+
+class ClientEthereumAddresses(NamedTuple):
+    provider: str
+    requestor: str
+
+
+class Report(NamedTuple):
+    pair_reports: list
+
+
+def create_report() -> Report:
+    """ Create payments report for Concent. """
+    pair_reports = []
+
+    for subtask in get_subtasks_with_distinct_clients_pairs():  # pylint: disable=not-an-iterable
+        (subtasks_ids, clients_ethereum_addresses) = get_subtask_ids_and_ethereum_addresses_for_clients_pair(
+            subtask.requestor,
+            subtask.provider,
+        )
+        related_pending_responses_for_pair = get_related_pending_responses_for_subtasks_ids(subtasks_ids, subtask.requestor, clients_ethereum_addresses)
+        try:
+            related_payments_for_pair = get_related_payments_for_pair(clients_ethereum_addresses)
+        except SCINotSynchronized:
+            related_payments_for_pair = Payments([], [], [], [], [])
+        is_related_payments_for_pair_empty = related_payments_for_pair.is_empty()  # pylint: disable=no-member
+        is_related_pending_responses_for_pair_empty = related_pending_responses_for_pair.is_empty()  # pylint: disable=no-member
+        if not(is_related_payments_for_pair_empty and is_related_pending_responses_for_pair_empty):  # pylint: disable=no-member
+            pair_report = PairReport(
+                requestor=subtask.requestor,
+                provider=subtask.provider,
+                requestor_ethereum_addresses=[client_ethereum_address.requestor for client_ethereum_address in clients_ethereum_addresses],
+                provider_ethereum_addresses=[client_ethereum_address.provider for client_ethereum_address in clients_ethereum_addresses],
+                pending_responses=related_pending_responses_for_pair,
+                payments=related_payments_for_pair,
+            )
+            pair_reports.append(pair_report)
+
+    return Report(pair_reports)
+
+
+def get_subtasks_with_distinct_clients_pairs() -> QuerySet:
+    """ Get Subtasks with distinct pairs of requestor and provider Clients. """
+    return Subtask.objects.order_by('provider', 'requestor').distinct('provider', 'requestor')
+
+
+def get_subtask_ids_and_ethereum_addresses_for_clients_pair(requestor: Client, provider: Client) -> Tuple:
+    """ Get Subtasks Ids and all Ethereum keys for Clients pair. """
+    subtask_ids = []
+    clients_ethereum_addresses = []
+
+    for subtask in Subtask.objects.filter(requestor=requestor, provider=provider):  # pylint: disable=no-member
+        subtask_ids.append(subtask.subtask_id)
+        task_to_compute = deserialize_message(subtask.task_to_compute.data.tobytes())
```

Comment:
What if GM version will not match here and deserialization will fail?

Reply:
Good point, we should add filtering by protocol_version

Reply:
This is actually a little bit more complicated. We ignored this problem in the first version of report on purpose, and we might continue doing that here too, depending on @cameel decision.

The problem is that if you will ignore it, it might simply fail with the first ever run.
The problem is that if you will not ignore it, and add filtering by procol version, the report output can be misleading, because later payments could not be matched with responses, and the calculated values will be different than real.

I think the simplest solution is to create report only for time period since latest GM version is being used. But that is bigger architecture decision.

Reply:
I agree with you. Let's wait for @cameel and then I will change or leave it

Reply:
See https://github.com/golemfactory/concent/issues/1044#issuecomment-453200820

Add all the information you need to `Subtask` or create models corresponding to the messages.

Please also make sure that your changes are backwards compatible with the current database schema.